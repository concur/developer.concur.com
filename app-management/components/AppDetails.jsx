/* eslint-env browser */

import React, { PropTypes } from 'react';
import { Link } from 'react-router';

class AppDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      app: {},
      loading: false,
      error: false,
    };
  }

  componentWillMount() {
    window.fetch(`${process.env.API_SERVER}/apps/${this.props.params.id}`)
      .then(response => response.json())
      .then(app => this.setState({
        app,
        loading: false,
      }))
      .catch(() => this.setState({
        error: true,
        loading: false,
      }));
  }

  render () {
    if (this.state.loading) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>Loading...</p>
          </div>
        </div>
      );
    }

    if (this.state.error) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>An error occurred when loading this app</p>
            <Link to="/">Back to My Apps</Link>
          </div>
        </div>
      );
    }

    const { app } = this.state;

    return (
      <div className="row">
        <div className="col-md-12">
          <h2>{app.name}</h2>
          <Link to="/">&larr; Back to My Apps</Link>
          <div className="row">
            <p>{app.description}</p>
          </div>
        </div>
      </div>
    );
  }
}

AppDetails.propTypes = {
  params: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }),
};

export default AppDetails;
