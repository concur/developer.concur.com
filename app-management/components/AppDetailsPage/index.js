/* eslint-env browser */

import { connect } from 'react-redux';
import AppDetailsPage from './AppDetailsPage';
import { fetchAppDetails, updateAppDetails, showSecret } from '../../actions/appDetails';

export function mapStateToProps(state) {
  return state.appDetails;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppDetails: id => dispatch(fetchAppDetails(id)),
    handleSubmit: app => dispatch(updateAppDetails(app)),
    generateSecret: (id) => {
      const confirmation = window.confirm('Are you sure? This will clear your previous app secret.');
      if (confirmation) {
        dispatch(showSecret());
        console.log(`TBD: Generate secret for app ${id}`);
      }
    },
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppDetailsPage);
