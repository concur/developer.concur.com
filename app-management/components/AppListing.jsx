import React from 'react';
import { Link } from 'react-router';

class AppListing extends React.Component {
  constructor(props) {
    super(props);
    const apps = [
      {
        id: '2342asdfas',
        name: 'Starbucks',
      },
      {
        id: '234asdfasdf',
        name: 'Uber',
      },
      {
        id: 'asldfkajsdflj',
        name: 'My App',
      },
    ];
    this.state = { apps };
  }

  render () {
    const appItems = this.state.apps.map((app) => {
      const { id, name } = app;
      return (
        <div className="well" key={id}>
          <h2>{name}</h2>
          <Link to="" className="btn-u btn-u-green">Details</Link>
          <Link to="" className="btn-u btn-u-blue">Edit</Link>
        </div>
      );
    });

    return (
      <div className="row">
        <div className="col-md-12">
          <h1>My Apps</h1>
          <div className="row">
            {appItems}
          </div>
        </div>
      </div>
    );
  }
}

export default AppListing;
