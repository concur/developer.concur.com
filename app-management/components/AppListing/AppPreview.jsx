import React, { PropTypes } from 'react';
import { Link } from 'react-router';

import CertificationBadge from '../CertificationBadge';

const AppPreview = ({ app: { id, name, certified } }) => (
  <div className="well col-md-12" key={id}>
    <h3>
      <Link to={`/details/${id}`}>{name} </Link>
      <CertificationBadge certified={certified} />
    </h3>
  </div>
);

AppPreview.propTypes = {
  app: PropTypes.object.isRequired,
};

export default AppPreview;
