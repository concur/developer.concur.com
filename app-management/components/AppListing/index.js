import { connect } from 'react-redux';
import AppListing from './AppListing';
import { fetchAppListing } from '../../actions/appListing';

export function mapStateToProps(state) {
  return state.appListing;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppListing: () => dispatch(fetchAppListing()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppListing);
