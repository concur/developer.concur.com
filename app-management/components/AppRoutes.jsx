import React, { PropTypes } from 'react';
import { Router, Route, IndexRoute, hashHistory } from 'react-router';

import PageContainer from './PageContainer';
import AppListing from './AppListing';
import AppDetailsPage from './AppDetailsPage';
import NewAppPage from './NewAppPage';
import LoginPage from './LoginPage';
import SignupPage from './SignupPage';
import LogoutPage from './LogoutPage';
import NotFound from './NotFound';

import { logout } from '../actions/auth';

class AppRoutes extends React.Component {
  constructor(props) {
    super(props);

    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.requireAuth = this.requireAuth.bind(this);
    this.showAuthForms = this.showAuthForms.bind(this);
    this.logout = this.logout.bind(this);
  }

  isAuthenticated() {
    const { store } = this.props;
    return !!store.getState().auth.token;
  }

  requireAuth(nextState, replaceState) {
    if (!this.isAuthenticated()) {
      replaceState('/login');
    }
  }

  showAuthForms(nextState, replaceState) {
    if (this.isAuthenticated()) {
      replaceState('/');
    }
  }

  logout(nextState, replaceState) {
    if (this.isAuthenticated()) {
      this.props.store.dispatch(logout());
    } else {
      replaceState('/login');
    }
  }

  render () {
    return (
      <Router history={hashHistory}>
        <Route path="/" component={PageContainer}>
          <IndexRoute component={AppListing} onEnter={this.requireAuth} />
          <Route path="new" component={NewAppPage} onEnter={this.requireAuth} />
          <Route path="details/:id" component={AppDetailsPage} onEnter={this.requireAuth} />
          <Route path="login" component={LoginPage} onEnter={this.showAuthForms} />
          <Route path="signup" component={SignupPage} onEnter={this.showAuthForms} />
          <Route path="logout" component={LogoutPage} onEnter={this.logout} />
          <Route path="*" component={NotFound} />
        </Route>
      </Router>
    );
  }
}

AppRoutes.propTypes = {
  store: PropTypes.shape({
    dispatch: PropTypes.func.isRequired,
  }),
};

export default AppRoutes;
