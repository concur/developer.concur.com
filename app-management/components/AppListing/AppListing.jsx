import React, { PropTypes } from 'react';
import LoadingSpinner from '../LoadingSpinner';
import ErrorAlert from '../ErrorAlert';
import AppPreview from './AppPreview';
import NewAppInstructions from '../NewAppPage/NewAppInstructions';

class AppListing extends React.Component {
  componentWillMount() {
    this.props.fetchAppListing();
  }

  render () {
    const { isFetching, error, apps } = this.props;
    let content;

    if (isFetching) {
      content = <LoadingSpinner loading={isFetching} />;
    } else if (apps.length > 0) {
      content = apps.map(app => <AppPreview app={app} key={app.id} />);
    } else {
      content = <NewAppInstructions />;
    }

    return (
      <div className="row">
        <div className="col-md-12">
          <h2><i className="fa fa-list" aria-hidden="true" /> My Apps</h2>
          <ErrorAlert error={error} />
          {content}
        </div>
      </div>
    );
  }
}

AppListing.propTypes = {
  fetchAppListing: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  error: PropTypes.string.isRequired,
  apps: PropTypes.arrayOf(PropTypes.shape()),
};

export default AppListing;
