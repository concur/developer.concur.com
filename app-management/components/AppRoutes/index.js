import AppRoutes from './AppRoutes';

export default AppRoutes;
