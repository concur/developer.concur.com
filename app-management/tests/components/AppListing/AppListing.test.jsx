import React from 'react';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import { createStore } from 'redux';
import { shallowToJson } from 'enzyme-to-json';

import AppListing from '../../../components/AppListing/AppListing';
import { mapStateToProps, mapDispatchToProps } from '../../../components/AppListing';
import AppPreview from '../../../components/AppListing/AppPreview';
import appReducer from '../../../reducers';
import appFactory from '../../app.mock';

const store = createStore(appReducer);

const defaultProps = {
  fetchAppListing: jest.fn(),
  isFetching: false,
  error: '',
  apps: [],
};

describe('<AppListing />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('calls componentWillMount to fetch the app listing', () => {
    sinon.spy(AppListing.prototype, 'componentWillMount');
    const listing = shallow(<AppListing {...defaultProps} />);

    expect(AppListing.prototype.componentWillMount.calledOnce).toBeTruthy();
    expect(defaultProps.fetchAppListing).toHaveBeenCalled();
  });

  it('renders apps', () => {
    const apps = [
      appFactory('id-1'),
      appFactory('id-2'),
    ];
    const props = {
      ...defaultProps,
      apps,
    };
    const listing = shallow(<AppListing {...props} />);

    expect(shallowToJson(listing)).toMatchSnapshot();
  });

  it('displays the loading spinner if the listing is loading', () => {
    const props = {
      ...defaultProps,
      isFetching: true,
    };
    const listing = shallow(<AppListing {...props} />);

    expect(shallowToJson(listing)).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('returns entire appListing state', () => {
      const state = mapStateToProps(store.getState());

      expect(state).toEqual(store.getState().appListing);
    });
  });

  describe('mapDispatchToProps', () => {
    it('returns submit handler', () => {
      const dispatcher = jest.fn();
      const state = mapDispatchToProps(dispatcher);
      state.fetchAppListing();

      expect(state.fetchAppListing).toBeDefined();
      expect(dispatcher).toHaveBeenCalled();
    });
  });
});
