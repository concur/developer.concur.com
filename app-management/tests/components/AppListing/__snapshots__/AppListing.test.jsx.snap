exports[`<AppListing /> displays the loading spinner if the listing is loading 1`] = `
<div
  className="row">
  <div
    className="col-md-12">
    <h2>
      <i
        aria-hidden="true"
        className="fa fa-list" />
       My Apps
    </h2>
    <ErrorAlert
      error="" />
    <LoadingSpinner
      loading={true} />
  </div>
</div>
`;

exports[`<AppListing /> renders apps 1`] = `
<div
  className="row">
  <div
    className="col-md-12">
    <h2>
      <i
        aria-hidden="true"
        className="fa fa-list" />
       My Apps
    </h2>
    <ErrorAlert
      error="" />
    <AppPreview
      app={
        Object {
          "allowedGrants": Array [
            "implicit",
            "authorization_code",
          ],
          "allowedScopes": Array [
            "receipts.read",
            "receipts.write",
          ],
          "appType": "Business",
          "certified": false,
          "description": "This is my app.",
          "enabled": true,
          "id": "id-1",
          "name": "Starbucks",
          "owner": "someidoftheowner",
          "redirectUris": Array [
            "http://localhost:3000",
            "http://www.concur.com",
          ],
        }
      } />
    <AppPreview
      app={
        Object {
          "allowedGrants": Array [
            "implicit",
            "authorization_code",
          ],
          "allowedScopes": Array [
            "receipts.read",
            "receipts.write",
          ],
          "appType": "Business",
          "certified": false,
          "description": "This is my app.",
          "enabled": true,
          "id": "id-2",
          "name": "Starbucks",
          "owner": "someidoftheowner",
          "redirectUris": Array [
            "http://localhost:3000",
            "http://www.concur.com",
          ],
        }
      } />
  </div>
</div>
`;
